service/history/replication/stream_sender_flow_controller.go

```diff
@@ -0,0 +1,135 @@
+		RefreshReceiverFlowControlInfo(syncState *replicationpb.SyncReplicationState)
+	}
+	SenderFlowControllerImpl struct {
+		flowControlStates  map[enums.TaskPriority]*flowControlState
+		defaultRateLimiter quotas.RateLimiter
+		logger             log.Logger
+	}
+)
+
+func NewSenderFlowController(config *configs.Config, logger log.Logger) *SenderFlowControllerImpl {
+	flowControlStates := make(map[enums.TaskPriority]*flowControlState)
+	highPriorityState := &flowControlState{
+		resume: true,
+	}
+	highPriorityState.cond = sync.NewCond(&highPriorityState.mu)
+	highPriorityState.rateLimiter = quotas.NewDynamicRateLimiter(quotas.NewDefaultOutgoingRateLimiter(func() float64 {
+		return float64(config.ReplicationStreamSenderHighPriorityMaxQPS())
+	}), 1*time.Minute)
+
+	lowPriorityState := &flowControlState{
+		resume: true,
+	}
+	lowPriorityState.cond = sync.NewCond(&lowPriorityState.mu)
+	lowPriorityState.rateLimiter = quotas.NewDynamicRateLimiter(quotas.NewDefaultOutgoingRateLimiter(func() float64 {
+		return float64(config.ReplicationStreamSenderLowPriorityMaxQPS())
+	}), 1*time.Minute)
+	flowControlStates[enums.TASK_PRIORITY_HIGH] = highPriorityState
+	flowControlStates[enums.TASK_PRIORITY_LOW] = lowPriorityState
+	return &SenderFlowControllerImpl{
+		flowControlStates:  flowControlStates,
+		defaultRateLimiter: quotas.NewRateLimiter(float64(config.ReplicationStreamSenderHighPriorityMaxQPS()), config.ReplicationStreamSenderHighPriorityMaxQPS()),
+		logger:             logger,
+	}
+}
+
+func (s *SenderFlowControllerImpl) RefreshReceiverFlowControlInfo(syncState *replicationpb.SyncReplicationState) {
+	if syncState.GetHighPriorityState() != nil {
+		s.setState(s.flowControlStates[enums.TASK_PRIORITY_HIGH], syncState.GetHighPriorityState().GetFlowControlCommand())
+	}
+	if syncState.GetLowPriorityState() != nil {
+		s.setState(s.flowControlStates[enums.TASK_PRIORITY_LOW], syncState.GetLowPriorityState().GetFlowControlCommand())
+	}
+}
+
+func (s *SenderFlowControllerImpl) setState(state *flowControlState, flowControlCommand enums.ReplicationFlowControlCommand) {
+	switch flowControlCommand {
+	case enums.REPLICATION_FLOW_CONTROL_COMMAND_RESUME, enums.REPLICATION_FLOW_CONTROL_COMMAND_UNSPECIFIED:
+		state.mu.Lock()
+		defer state.mu.Unlock()
+		state.resume = true
+		if state.waiters > 0 {
+			state.cond.Signal()
+		}
+	case enums.REPLICATION_FLOW_CONTROL_COMMAND_PAUSE:
+		state.mu.Lock()
+		defer state.mu.Unlock()
+		state.resume = false
+	}
+}
+
+func (s *SenderFlowControllerImpl) Wait(priority enums.TaskPriority) {
+	state, ok := s.flowControlStates[priority]
+	waitForRateLimiter := func(rateLimiter quotas.RateLimiter) {
+		err := rateLimiter.Wait(context.Background())
+		if err != nil {
+			s.logger.Error("error waiting for rate limiter", tag.Error(err))
+		}
+		return
+	}
+	if !ok {
+		waitForRateLimiter(s.defaultRateLimiter)
+		return
+	}
+	state.mu.Lock()
+	defer state.mu.Unlock()
+	if state.resume {
+		waitForRateLimiter(state.rateLimiter)
```

Comment:
could it cause problem if it wait on the rate limiter while holding the lock?

Reply:
Nice catch. It would, it will block the sender receiver event loop. Let me fix it.